daemon/cluster/listen_addr.go

```diff
@@ -0,0 +1,160 @@
+package cluster
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+)
+
+func resolveListenAddr(specifiedInterface, specifiedAddr string) (string, error) {
+	specifiedHost, specifiedPort, err := net.SplitHostPort(specifiedAddr)
+	if err != nil {
+		return "", fmt.Errorf("could not parse listen address %s", specifiedAddr)
+	}
+	ip := net.ParseIP(specifiedHost)
+	if ip != nil && ip.IsUnspecified() {
+		// The user has not provided a listen address.
+		if specifiedInterface != "" {
+			interfaceComponents := strings.SplitN(specifiedInterface, "/", 2)
+			if len(interfaceComponents) > 1 {
+				specifiedPort = interfaceComponents[1]
+				_, err := strconv.ParseUint(specifiedPort, 10, 16)
+				if err != nil {
+					return "", fmt.Errorf("could not parse listen interface %s", specifiedInterface)
+				}
+			}
+			interfaceAddr, err := resolveInterfaceAddr(interfaceComponents[0])
+			if err != nil {
+				return "", err
+			}
+			return net.JoinHostPort(interfaceAddr.String(), specifiedPort), nil
+		}
+
+		systemAddr, err := resolveSystemAddr()
+		if err != nil {
+			return "", err
+		}
+		return net.JoinHostPort(systemAddr.String(), specifiedPort), nil
+	} else if specifiedInterface != "" {
+		return "", errors.New("cannot specify both --listen-interface and --listen-addr")
+	}
+
+	return specifiedAddr, nil
+}
+
+func resolveInterfaceAddr(specifiedInterface string) (net.IP, error) {
+	// Use a specific interface's IP address.
+	intf, err := net.InterfaceByName(specifiedInterface)
+	if err != nil {
+		return nil, err
+	}
+
+	addrs, err := intf.Addrs()
+	if err != nil {
+		return nil, err
+	}
+
+	var interfaceAddr4, interfaceAddr6 net.IP
+
+	for _, addr := range addrs {
+		ipAddr, ok := addr.(*net.IPNet)
+
+		if ok {
+			if ipAddr.IP.To4() != nil {
+				// IPv4
+				if interfaceAddr4 != nil {
+					return nil, fmt.Errorf("interface %s has more than one IPv4 address", specifiedInterface)
+				}
+				interfaceAddr4 = ipAddr.IP
+			} else {
+				// IPv6
+				if interfaceAddr6 != nil {
+					return nil, fmt.Errorf("interface %s has more than one IPv6 address", specifiedInterface)
+				}
+				interfaceAddr6 = ipAddr.IP
+			}
+		}
+	}
+
+	if interfaceAddr4 == nil && interfaceAddr6 == nil {
+		return nil, fmt.Errorf("interface %s has no usable IPv4 or IPv6 address", specifiedInterface)
+	}
+
+	// In the case that there's exactly one IPv4 address
+	// and exactly one IPv6 address, favor IPv4 over IPv6.
+	if interfaceAddr4 != nil {
+		return interfaceAddr4, nil
+	}
+	return interfaceAddr6, nil
+}
+
+var errMultipleIPs = errors.New("could not choose a listening IP address since this system has multiple addresses - specify one with --listen-addr or --listen-interface")
```

Comment:
Not today, nothing like errcode or anything.
For 1.12 we are going to be sending json errors at least, but it only has a `message` field.
We'll have to check strings for 1.12, I think.